Frameworks/CoreGraphics/CGPath.mm

```diff
@@ -244,126 +272,107 @@ void CGPathAddLineToPoint(CGMutablePathRef path, const CGAffineTransform* transf
 
     path->BeginFigure();
     path->GetGeometrySink()->AddLine(_CGPointToD2D_F(pt));
+    path->SetLastTransform(transform);
 
     path->SetCurrentPoint(pt);
 }
 
-CGFloat _CGPathControlPointOffsetMultiplier(CGFloat angle) {
-    // Constant used to approximate circles with bezier curves.
-    // An n-piece cubic Bezier curve can approximate a circle,
-    // when each inner control point is the distance 4/3 * tan(t/4)
-    // from an outer control point on a unit circle, where t is 360/n degrees, and
-    // n > 2
-    return (4.0f / 3.0f) * tan(angle / 4.0f);
+static inline CGPoint _getInvertedCurrentPointOfPath(CGPathRef path) {
+    CGPoint point = path->GetCurrentPoint();
+    if (!CGAffineTransformEqualToTransform(*path->GetLastTransform(), CGAffineTransformIdentity)) {
+        point = CGPointApplyAffineTransform(point, CGAffineTransformInvert(*path->GetLastTransform()));
+    }
+    return point;
+}
+
+static HRESULT _createPathReadyForFigure(CGPathRef previousPath,
+                                         CGPoint startPoint,
+                                         ID2D1PathGeometry** pathGeometry,
+                                         ID2D1GeometrySink** geometrySink) {
+    ComPtr<ID2D1Factory> factory;
+    RETURN_IF_FAILED(_CGGetD2DFactory(&factory));
+    RETURN_IF_FAILED(factory->CreatePathGeometry(pathGeometry));
+    RETURN_IF_FAILED((*pathGeometry)->Open(geometrySink));
+
+    CGPoint invertedPoint = _getInvertedCurrentPointOfPath(previousPath);
+    if (!CGPointEqualToPoint(invertedPoint, startPoint)) {
+        (*geometrySink)->BeginFigure(_CGPointToD2D_F(invertedPoint), D2D1_FIGURE_BEGIN_FILLED);
+        (*geometrySink)->AddLine(_CGPointToD2D_F(startPoint));
```

Comment:
nit: I am not comfortable with using this indirect throughout. We could save these in locals and copyto or release them to store them into the out pointers.